src/common/operator/cast_operators.cpp

```diff
@@ -1542,7 +1546,27 @@ struct DecimalCastOperation {
 
 	template <class T, bool NEGATIVE>
 	static bool HandleExponent(T &state, int32_t exponent) {
-		Finalize<T>(state);
+		state.exponent = true;
+		auto decimal_excess = (state.decimal_count > state.scale) ? state.decimal_count - state.scale : 0;
+		if (exponent > 0) {
+			//! Positive exponents need up to 'exponent' amount of digits
+			//! Everything beyond that amount needs to be truncated
+			if (decimal_excess > exponent) {
+				//! We've allowed too many decimals
+				state.excessive_decimals = decimal_excess - exponent;
+				exponent = 0;
+			} else {
+				exponent -= decimal_excess;
+			}
+			D_ASSERT(exponent >= 0);
+		} else if (exponent < 0 && decimal_excess) {
+			//! Negative exponents dont require any extra digits
+			state.excessive_decimals = -exponent + decimal_excess;
+			exponent = 0;
```

Comment:
This somehow looks wrong?
As soon as we have *any* excess decimals, the negative exponent is completely ignored?

Reply:
This is not wrong, because TruncateExcessiveDecimals is almost identical to the logic that happens to negative exponents, adding it to 'excessive_decimals' makes the negative exponent logic obsolete in this case.

In the new (hopefully last) update TruncateExcessiveDecimals in this case is performed as if no exponent was encountered, and the exponent is carried out after the Finalize step.